### ISSUE TITLE
Difficulty using normalizedTransform with presetFixedRatioType

### ISSUE THREAD

**Issue by luke-oneill** [guoyingtao/Mantis#97]
I want to open a crop view using crop data that is fetched from a server in the format `x, y, width, height`.  Using the following code to initialise with a normalizedTranform:

```
            let normalisedTransform = CGRect(
                x: crop.minX / image.size.width,
                y: crop.minY / image.size.height,
                width: crop.width / image.size.width,
                height: crop.height / image.size.height
            )

            config.presetTransformationType = .presetNormalizedInfo(normailizedInfo: normalisedTransform)
```

**This works.**

Additionally, I want to constrain the crop to a particular ratio. Add the following line:

```
      config.presetFixedRatioType = .alwaysUsingOnePresetFixedRatio(ratio: cropRatio)
```

**This breaks.**

The transformation when crop view is loaded incorrectly – it is much more zoomed in than expected.

I'm not sure if I need to change my normalised transformation to take into account the ratio? Or is it a bug, or unsupported feature? Appreciate some guidance – happy to raise a PR if needed :) 

Let me know if you need any more info.

**Comment by guoyingtao** [comment]
I didn't test the case of combining normalizedTranform and .alwaysUsingOnePresetFixedRatio, so it probably does not support this feature very well.
You are welcome to submit a PR for it. It may be helpful if you check the code in **viewDidAppear** in **CropViewController**

One question is when you want to constrain the crop to a particular ratio, do you also want to  constrain the crop inside the rect after applying normalizedTranform?

**Comment by luke-oneill** [comment]
I've done some digging, and found a couple of things:

1. The problem only occurs when the preset ratio is less than the image's aspect ratio. e.g. using the following code in the example project:

```
        var config = Mantis.Config()
        
        let transform = CGRect(x: 0, y: 0, width: 1, height: 1)
        config.presetTransformationType = .presetNormalizedInfo(normailizedInfo: transform)
        
        let cropViewController = Mantis.cropViewController(image: image,
                                                           config: config)
        cropViewController.modalPresentationStyle = .fullScreen
        cropViewController.delegate = self
        cropViewController.config.presetFixedRatioType = .alwaysUsingOnePresetFixedRatio(ratio: RATIO)
        present(cropViewController, animated: true)
        
```
 where RATIO is 5:

![horizontal](https://user-images.githubusercontent.com/5530491/114745569-aff0d680-9d46-11eb-94a5-1c659175c717.jpg)

vs where RATIO is 0.2:

![vertical](https://user-images.githubusercontent.com/5530491/114745695-d44cb300-9d46-11eb-9426-d1a1db87b153.jpg)

2. Changing `setCropBoxFrame(...)` in `CropViewModel.swift` to change the if/else statement to the single line:
```
 cropBoxFrame.size.height = cropBoxFrame.width / aspectRatio
```
... will resolve the issue. But it creates at least one new problem: images with a preset ratio that is less than the image aspect, and without a transform will load incorrectly:

<img width="116" alt="Screenshot 2021-04-14 at 17 20 08" src="https://user-images.githubusercontent.com/5530491/114746116-491fed00-9d47-11eb-9f29-45199e653590.png">

My guess is I need work on the `setCropBoxFrame(...)` function more so it works in both situations, but I'm stuck at this point. Do you have any thoughts or pointers?

**Comment by guoyingtao** [comment]
@luke-oneill 

I found a way to support this feature. I am using the code below, RATIO is 0.2 or 5. I attached a result screenshot here, just want to make sure it is what you want.

```Swift
var config = Mantis.Config()
        
        let transform = CGRect(x: 0.1, y: 0.1, width: 0.8, height: 0.8)
        config.presetTransformationType = .presetNormalizedInfo(normailizedInfo: transform)
        
        let cropViewController = Mantis.cropViewController(image: image,
                                                           config: config)
        cropViewController.modalPresentationStyle = .fullScreen
        cropViewController.delegate = self
        cropViewController.config.presetFixedRatioType = .alwaysUsingOnePresetFixedRatio(ratio: RATIO)
        present(cropViewController, animated: true)
```

![image](https://user-images.githubusercontent.com/26723384/114807794-06750980-9d75-11eb-8e1a-b7c6f91b9390.png)

**Comment by luke-oneill** [comment]
Awesome, looks great! Thank you so much.

One question is how do the transform values work? I would assume a width and height of 0.8 means the crop is not the full size of the image, but in image the crops go the full width or height of the image.

**Comment by guoyingtao** [comment]
@luke-oneill The crops get 0.8 width and 0.8 height.
The current logic is the image always zooms into the crop area.
![image](https://user-images.githubusercontent.com/26723384/114871332-30571c00-9dc7-11eb-938b-2a599f3aede9.png)

**Comment by luke-oneill** [comment]
Ahah, yeah I see now, makes perfect sense and solves my use case. Thanks!

**Comment by guoyingtao** [comment]
Good! I will add the logic to a new branch later for you to test.
Once it passes your tests, I will release a new version.

**Comment by guoyingtao** [comment]
Can you try the pod settings below to see if it solved your issue?

```Ruby
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => '4a3ead5c3c431aaae1e143d386fa5d6f99861d9d'
```

**Comment by luke-oneill** [comment]
Tested, and it works perfectly 🥳

**Comment by luke-oneill** [comment]
Think I found an issue actually. 

For my test image and ratio, the y transform works as expected, but the x transform is off by the width of the crop box. See the image below, I've got four different transforms – I'd expect the crops to be all the way to the left for x = 0, and all the way to the right for x = 1. 

![IMG_0005](https://user-images.githubusercontent.com/5530491/115000457-b001eb00-9e9a-11eb-896f-3ab642306f6a.JPG)

**Comment by guoyingtao** [comment]
Since it is normalized info, it should have the limitation like 0 <= (x + width) <=1, 0 <= (y + height) < = 1. So I think for your case you should use (0,0,0.5,0.5), (0.5,0,0.5,0.5),(0,0.5,0.5,0.5),(0.5,0,.5,0.5,0.5)
Also, .alwaysUsingOnePresetFixedRatio will always get a crop which will be in the middle of the crop after nomalizedTransform.

You can see the red and blue crop box in the pictures below

![](https://user-images.githubusercontent.com/26723384/114809469-3d98ea00-9d78-11eb-9172-62ffbf1c822e.png)

**Comment by luke-oneill** [comment]
I think I understand, thanks. Are you're saying that this specific use case is not possible right now:

**Open a crop view, with a fixed ratio, that is transformed to any point within the image, including the far edges.**

i.e. It is impossible to open a crop view with a crop that matches the yellow rectangles?

![IMG_0005](https://user-images.githubusercontent.com/5530491/115255547-4990fd00-a126-11eb-8020-69ade24a8676.JPG)

![Screenshot 2021-04-19 at 15 46 51](https://user-images.githubusercontent.com/5530491/115255775-80671300-a126-11eb-8eb5-78501400bcfb.png)

**Comment by guoyingtao** [comment]
@luke-oneill We may need extra parameters for this scenario.
For example, we can add a new presetFixRatioType
```Swift
public enum PresetFixRatioPositionType {
  case middle
  case topLeft (top for ratio is some values like 5 and left for ratio is some values like 0.2)
  case topRight
  case bottomLeft
  case bottomRight
}

public enum PresetFixedRatioType {
    case alwaysUsingOnePresetFixedRatio(ratio: Double, type: PresetFixRatioPositionType = .middle)
    case canUseMultiplePresetFixedRatio(defaultRatio: Double = 0)
}
```

**Comment by luke-oneill** [comment]
I don't think that works... if I select `.topLeft` it would result in the yellow crops below. I want a user to be able to crop anywhere on the photo.

![IMG_0005](https://user-images.githubusercontent.com/5530491/115272677-94b30c00-a136-11eb-84b9-9cb9c569a24b.JPG)

I'm not sure I've made my exact use case 100% clear, so will try again:

1. Save a fixed ratio crop of anywhere on an image.
2. Store this information in a database.
3. Load a crop view (on a different physical device) which is set to the same crop that was previously saved, and enforces the same fixed ratio.

Maybe there is a simpler way of going about this that I'm missing?

**Comment by guoyingtao** [comment]
@luke-oneill Do you load the crop view on the same image?

**Comment by luke-oneill** [comment]
Yep, same image.

**Comment by guoyingtao** [comment]
@luke-oneill I made some changes in the code. Now you can try using  config.presetTransformationType = .presetInfo(info: transform)
You can retrieve transform information from 
```Swift
cropViewControllerDidCrop(_ cropViewController: CropViewController, cropped: UIImage, transformation: Transformation)
```
then save it to your server

If you think it works then you can try 
```Ruby
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => 'bb9bfd70cad5ec27b0545ddef007533bfb3f8678'
```

**Comment by guoyingtao** [comment]
@luke-oneill I made some changes in the code. Now you can try using  config.presetTransformationType = .presetInfo(info: transform)
You can retrieve transform information from 
```Swift
cropViewControllerDidCrop(_ cropViewController: CropViewController, cropped: UIImage, transformation: Transformation)
```
then save it to your server

If you think it works then you can try 
```Ruby
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => 'bab58bad54bd4323985e273450140054665e5050'
```

**Comment by guoyingtao** [comment]
@luke-oneill I made some changes in the code. Now you can try using  config.presetTransformationType = .presetInfo(info: transform)
You can retrieve transform information from 
```Swift
cropViewControllerDidCrop(_ cropViewController: CropViewController, cropped: UIImage, transformation: Transformation)
```
then save it to your server

If you think it works then you can try 
```Ruby
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => 'c3d1b2a164ae6d500fef3cc2cafe1c62eccd6eae'
```

Then use
```Swift
config.presetTransformationType = .presetInfo(info: transform)
config.presetFixedRatioType = .alwaysUsingOnePresetFixedRatio()
```

**Comment by luke-oneill** [comment]
Thanks, this is close! It works for saving/loading some crop rectangles, but on some it breaks – loading a crop outside the boundaries of the image, e.g.

Cropping the following rectangle (zoomed in to the centre bottom of the test image):

![1](https://user-images.githubusercontent.com/5530491/115355714-d5e90180-a1b2-11eb-86b2-8fdce60e7e35.jpg)

Gives the following transformation:

```
offset: (534.3333333333334, 409.0), rotation: 0.0, scale: 3.7149149887605, manualZoomed: true, intialMaskFrame: (73.33214285714287, 14.000000000000014, 243.33571428571426, 487.66666666666663), maskFrame: (73.54339622641511, 14.0, 242.91320754716978, 487.66666666666663), scrollBounds: (534.3333333333334, 409.0, 242.91320754716978, 487.66666666666663)
```

Which, if loaded on same or other devices, gives a crop as follows:

<img width="531" alt="Screenshot 2021-04-20 at 08 23 54" src="https://user-images.githubusercontent.com/5530491/115355966-1779ac80-a1b3-11eb-98be-2433c2a6de7e.png">


Follow up question: can I use the Mantis transformation values to derive a simple rectangle crop (x, y, width, height) relative to the original photo's size and coordinates?

**Comment by guoyingtao** [comment]
@luke-oneill Would you mind posting the image your are using here? Thank you!

**Comment by luke-oneill** [comment]
Sure :) Size is 3000 x 2002px (to double check GitHub din't resize!)

![IMG_0005](https://user-images.githubusercontent.com/5530491/115401849-bec20800-a1e2-11eb-9fc0-ebfdf40d5676.JPG)

**Comment by guoyingtao** [comment]
@luke-oneill I think I fixed the issue.
Can you try
```Swift
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => 'c3d1b2a164ae6d500fef3cc2cafe1c62eccd6eae'
```

**Comment by guoyingtao** [comment]
@luke-oneill I think I fixed the issue. Can you try

```Swift
pod 'Mantis', :git => 'https://github.com/guoyingtao/Mantis.git', :commit => '0d7b48a53beacc239aeb1924a91ec964cf4e006d'
```

**Comment by luke-oneill** [comment]
No change I'm afraid :(

BTW it's reproducible in your example project, if you select the above sample image from the simulators photo library, and set the transformation in `presentWithPresetTransformation` to:
```
        let transform = (
            offset: CGPoint(x: 534.3333333333334, y: 409.0),
            rotation: CGFloat(0),
            scale: CGFloat(3.7149149887605),
            manualZoomed: true,
            intialMaskFrame: CGRect(x: 73.33214285714287, y: 14.000000000000014, width: 243.33571428571426, height: 487.66666666666663),
            maskFrame: CGRect(x: 73.54339622641511, y: 14.0, width: 242.91320754716978, height: 487.66666666666663),
            scrollBounds: CGRect(x: 534.3333333333334, y: 409.0, width: 242.91320754716978, height: 487.66666666666663)
        )
```

**Comment by guoyingtao** [comment]
@luke-oneill  Can you check if the code below is changed. If not, can you delete the pod file and run `pod install` again?

<img width="752" alt="image" src="https://user-images.githubusercontent.com/26723384/115456684-7231f900-a1f1-11eb-9cac-1bee4128f400.png">

**Comment by luke-oneill** [comment]
Yep, all updated correctly. I'm using Swift Package Manager, but did the equivalent – deleted the checked out files and re-setup the dependency!

And to triple check I shut down Xcode and my simulators, and running it on a fresh simulator, so no code could have been cached.

**Comment by luke-oneill** [comment]
Ohh! Hold on, this update changes the transformation that gets output? I'm sorry, let me check that way round – I was thinking you changed the code that loads and applies the transformation!

**Comment by guoyingtao** [comment]
The old code is `intialMaskFrame: viewModel.cropOrignFrame` and the new one is `getInitialCropBoxRect()`

**Comment by guoyingtao** [comment]
Yes, the update changes he transformation.

**Comment by luke-oneill** [comment]
Sorry my bad – tested again properly and you're right, this fixes it :)
I'll wire it in properly to my server load/save in the morning so I can test it more thoroughly, but it looks good for now. Thanks so much for your support with this!

The final piece of the puzzle is I need to be compatible with crops created from a web interface (i.e. not Mantis). I would like to convert to a simple `x, y, width, height` crop, in relation to the source image's coordinates. I can successfully convert from a Mantis Transformation to that format with this code:

```
        let crop = CGRect(
            x: CGFloat(transformation.scrollBounds.minX / (transformation.scrollBounds.width / cropped.size.width)),
            y: CGFloat(transformation.scrollBounds.minY / (transformation.scrollBounds.height / cropped.size.height)),
            width: cropped.size.width,
            height: cropped.size.height
        )
```

but I haven't figured out how to go the other way. Do you know if that is possible?

**Comment by guoyingtao** [comment]
@luke-oneill If you don't need to rotate the image, I think the best way to do it is still using presetNormalizedInfo(normailizedInfo: CGRect).
I changed the code so if you use config.presetFixedRatioType = .alwaysUsingOnePresetFixedRatio(), it will keep the ratio of preset transformation.
Also it is easier to convert between Mantis and your web interface.

**Comment by guoyingtao** [comment]
@luke-oneill I released Mantis 1.6.1 which fixed a couple of issues. Can you check if you can use presetNormalizedInfo(normailizedInfo: CGRect). to solve this issue then?

**Comment by luke-oneill** [comment]
Hey, all tested and works well. Thanks so much for this release!
